CPP: Ethernet devices plus connection strings.

michaelpantic opened the pull request:
- Added TCP devices using regular linux sockets
- Changed factory such that device types can be configured with a sort-of connection string.
Examples for Parameter device_ids:

  - `usb://*` - Adds all Piksi devices on the USB bus. This is the default.
  - `usb://12345` - Adds piksi with serial number 12345 on USB
  - `tcp://192.168.0.222:55555` - Adds piksi at IP ...222 and default tcp port
  - `usb://*;tcp://192.168.0.222:55555`  - Adds all piksis on USB and the one on the network with this ip.

   - ( they can be combined arbitrarily).

- Todo in another PR: Serial devices. They will also get their connection string (`serial:///dev/ttyUSB0` etc).

rikba commented:
Adresses ethernet interface requested in https://github.com/ethz-asl/ethz_piksi_ros/issues/102

rikba commented on piksi_multi_cpp/src/device/device_tcp.cc:
```diff
@@ -0,0 +1,100 @@
+  // set receive timeout to 10s such that read methods blocks not forever when
+  // something goes wrong
+  struct timeval tv;
+  tv.tv_sec = 10; /* 10 Secs Timeout */
```
Nice. Should implement something like this for serial connections.

michaelpantic replied:
done in the serial port driver ;-)

rikba commented on piksi_multi_cpp/src/device/device_tcp.cc:
```diff
@@ -0,0 +1,100 @@
+  } else {
+    ROS_WARN_STREAM("TCP error " << received_length << " while reading device "
+                                 << id_);
+    return 0;
```
In the USB device we return the read value immediately and let [libsbp do the error handling](https://github.com/swift-nav/libsbp/blob/master/c/src/sbp.c#L392-L431). But probably we should also do error handling in the other devices.

michaelpantic replied:
yeah don't know, here its a bit special because recvfrom can return negative stuff and I wanted to catch this ;-)

michaelpantic commented on piksi_multi_cpp/src/device/device_factory.cc:
```diff
@@ -1,31 +1,67 @@
+ */
+std::vector<Device::Ptr> DeviceFactory::createByIdentifier(
+    const piksi_multi_cpp::Identifier& id) {
+  std::regex rgx("^(.+):\\/\\/(.+)$");  // check if id is of type protocol://id
```
https://xkcd.com/208/

michaelpantic commented:
@rikba addressed comments, thanks for the review! I think CLion sneaks these quotationmarks in when it adds references automatically ... should change this ;-)
